### DISCUSSION TITLE
Not understanding setFreeFont, nonsense display

### DISCUSSION THREAD

**Discussion by degtyarev708** [Bodmer/TFT_eSPI#1076] [Q&A - Fonts] [answered]
So Im misunderstanding _something_ here, and any help would be greatly appreciated. Im using an stm32f103 with an ST7735 128*160 display. The display works fine, both drawString and print work fine. After I set the font to one of those in GFXFF however, it all goes off the rails. Fills the entire display with the background with text color artifacts both inside and outside the bounds. Different fonts also give different artifacts, as does changing the drawString position.

Picture is a bit rough, but you can see that its not exactly the expected output
![20210315_023847](https://user-images.githubusercontent.com/6896936/111114283-68a1db00-8539-11eb-8278-d86a3ed6c16a.jpg)

platformIO.ini
```
[env:blackpill_f103c8]
platform = ststm32
board = blackpill_f103c8
framework = arduino
upload_protocol = stlink
build_flags = 
	-D PIO_FRAMEWORK_ARDUINO_ENABLE_CDC
	-D USBCON
	-D USBD_VID=0x0483
	-D USBD_PID=0x5740
	-D USB_MANUFACTURER="Unknown"
	-D USB_PRODUCT="\"BLACKPILL_F103C8\""
	-DUSER_SETUP_LOADED=1
	-DST7735_DRIVER=1
	-DTFT_WIDTH=240
	-DTFT_HEIGHT=240
	-DTFT_MISO=PA6
	-DTFT_MOSI=PA7
	-DTFT_SCLK=PA5
	-DTFT_RST=PA4
	-DTFT_DC=PA3
	-DTFT_CS=PA2
	-DSPI_FREQUENCY=27000000
	-DSTM32=1
	-DST7735_GREENTAB=1
	-DLOAD_GLCD=1
	-DLOAD_GFXFF=1
lib_deps = 
	adafruit/Adafruit BusIO@^1.6.0
	adafruit/Adafruit GFX Library@^1.10.6
	bodmer/TFT_eSPI@^2.3.61`
```
Main.cpp
```c++
#include <Arduino.h>
#include <Wire.h>
#include <SPI.h>
//#include <EEPROM.h>
#include <Adafruit_GFX.h>
#include <TFT_eSPI.h>
#include <Fonts/GFXFF/FreeMonoBold9pt7b.h>
TFT_eSPI display = TFT_eSPI();
void setup() {
	display.init();
	display.fillScreen(TFT_BLUE);
	display.setTextColor(TFT_GREEN, TFT_MAGENTA);
	display.setCursor(10, 10);
	display.println("Default Font");
	display.drawString("drawString Default", 10, 20);
	display.setFreeFont(&FreeMonoBold9pt7b);
	display.drawString("drawString Font", 30, 30);
}

void loop() {
}
```

**Comment by Bodmer** [comment, marked as answer]
I think the compiler must be getting confused as it is not necessary to include the font or the Adafruit_GFX library. The fonts are pulled in automatically by the library.

Try this:
```
#include <Arduino.h>
#include <Wire.h>
#include <SPI.h>
// Not needed --- > #include <Adafruit_GFX.h>
#include <TFT_eSPI.h>
// Do not include ---> #include <Fonts/GFXFF/FreeMonoBold9pt7b.h>
TFT_eSPI display = TFT_eSPI();
void setup() {
  display.init();
  display.fillScreen(TFT_BLUE);
  display.setTextColor(TFT_GREEN, TFT_MAGENTA);
  display.setCursor(10, 10);
  display.println("Default Font");
  display.drawString("drawString Default", 10, 20);
  display.setFreeFont(&FreeMonoBold9pt7b);
  display.drawString("drawString Font", 30, 30);
}

void loop() {
}
```

**Reply by degtyarev708** [reply]
Doh, that was it. Wonderful, thank you!

**Reply by git2212** [reply]
Hi. Need to resurrect this very old thread as I ran into a similar issue... 

I got myself into the same situ by following what appear to be valid instructions in https://github.com/Bodmer/TFT_eSPI/blob/master/Fonts/GFXFF/gfxfont.h... which states "...To use a font in your Arduino sketch, #include the corresponding .h
// file and pass address of GFXfont struct to setFont()" ... ok setFreeFont(...) i figured.

Can you please elaborate on what you mean by the "compiler is getting confused"? It should not matter where the .h file is  being loaded from if the contents are the same... what am I missing?

**Reply by Bodmer** [reply]
The comment in gfxfont.h is misleading in this case, it was intended to relate to adding new customised fonts to the list in the file gfxfont.h. From the sketch viewpoint the comment can be ignored.

In the above case I suspect 3 copies of the font file was being loaded and the compiler then failed to link in a valid copy for some reason, so garbage values were then read from the FLASH memory during font rendering. It may be that the FLASH space was exceeded but that would be expected to generate a compile warning.

Can you describe your "similar issue" and test the code included in the selected answer above.

**Comment by git2212** [comment]
Well, long story short, I have the GFX font loaded as I need the font metrics for coordinate calculations, and tried to avoid your library loading it again, I tried passing my version of the font bitmap using your definition of GFXfont (and GFXglyph) to setFont()... i.e. 

typedef struct {
  uint8_t *bitmap;  ///< Glyph bitmaps, concatenated
  GFXglyph *glyph;  ///< Glyph array
  uint16_t first;   ///< ASCII extents (first char)
  uint16_t last;    ///< ASCII extents (last char)
  uint8_t yAdvance; ///< Newline distance (y axis)
} GFXfont;

...no joy for whatever reason as I get garbage on the screen.
if I could grab the GFXfont structure from your library then i could make the calculations based on the library version?

Cheers

**Reply by Bodmer** [reply]
The simplest way to do this is to make the font pointer acessible. i.e. move the following (which is protected currently) to be public:
https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.h#L952-L954

Then this will return the first code point value:
uint16_t first = pgm_read_word(&display.gfxFont->first);

For the existing fonts this will be 32 (space).

**Comment by git2212** [comment]
thanks for the quick reply.

Yep, this would certainly be a temporary solution as going down that path gives access to GFX font metrics and not anti aliased fonts?.... if I read the code correctly. Also, upgrading to the latest version of the lib would require the same hack lol.

The inherent complications notwithstanding, for my purposes, I just need the bounding rectangle that fully encloses a given ASCII string for a particular font that it would be rendered with. The ideal solution would not require repeated calls to set..Font()... 

perhaps that functionality is already there and I am missing it? 

BTW, awesome job on this lib :)

**Reply by Bodmer** [reply]
Are you still getting garbage on the screen?

**Comment by git2212** [comment]
Hi,

I can confirm that any font that is not loaded by your library will not work... and yes, i can get GFX font metrics as long as am using the ones loaded by your library. 

As I didn't like modifying your code, I ended up subclassing your TFT_eSPI class which gave me access to the protected members. I also ended up using the textWidth(...) and fontHeight(...) members of the TFT_eSPI class like this

   setFreeFont(gfxFont);
   float textSizeRatio=textsize_x/textsize;
  *w=this->textWidth(str)*textSizeRatio;
  *h=this->fontHeight()*textSizeRatio;

for the above to work i had to use setFreeFont but this proves less of an issue since I don't make assumptions about sticky fonts, settings during rendering... the above code completes in 20 micros in my RP2040 so there seems to be negligible performance penalties, mileage may vary for those running on slower processors ;)...

for the GFX fonts I've tested this solution seems to work for now.

Thx
